queue/group_remote_mongo_single.go

```diff
@@ -0,0 +1,226 @@
+package queue
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+type remoteMongoQueueGroupSingle struct {
+	canceler context.CancelFunc
+	client   *mongo.Client
+	mu       sync.RWMutex
+	opts     RemoteQueueGroupOptions
+	dbOpts   MongoDBOptions
+	queues   map[string]amboy.Queue
+}
+
+// NewMongoRemoteSingleQueueGroup constructs a new remote queue group. If ttl is 0, the queues will not be
+// TTLed except when the client explicitly calls Prune.
+func NewMongoRemoteSingleQueueGroup(ctx context.Context, opts RemoteQueueGroupOptions, client *mongo.Client, mdbopts MongoDBOptions) (amboy.QueueGroup, error) {
+	if err := opts.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid remote queue options")
+	}
+
+	if mdbopts.DB == "" {
+		return nil, errors.New("no database name specified")
+	}
+
+	if mdbopts.URI == "" {
+		return nil, errors.New("no mongodb uri specified")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	g := &remoteMongoQueueGroupSingle{
+		canceler: cancel,
+		client:   client,
+		dbOpts:   mdbopts,
+		opts:     opts,
+		queues:   map[string]amboy.Queue{},
+	}
+
+	if opts.PruneFrequency > 0 {
+		go func() {
+			pruneCtx, pruneCancel := context.WithCancel(context.Background())
+			defer pruneCancel()
+			defer recovery.LogStackTraceAndContinue("panic in remote queue group ticker")
+			ticker := time.NewTicker(opts.PruneFrequency)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					grip.Error(message.WrapError(g.Prune(pruneCtx), "problem pruning remote queue group database"))
+				}
+			}
+		}()
+	}
+
+	if opts.BackgroundCreateFrequency > 0 {
+		go func() {
+			defer recovery.LogStackTraceAndContinue("panic in remote queue group ticker")
```

Comment:
These defers and the one in the above goroutine are inconsistently ordered. I think the panic handler should be the first on the stack.